### ISSUE TITLE
Weird/Broken behaviour when there are more than 4 warnings

### ISSUE THREAD

**Issue by CastixGitHub** [flycheck/flycheck#1643]
## Checklist
- [x] I have checked existing issues for potential duplicates before creating this one.
- [x] I have read the [Troubleshooting guide][].

## Bug description
window splits and mode changes in both windows into `(Flycheck error messages FlyC-)` using flake8 (idk if limited to it) when there are more than 4 warnings. both 

## Steps to reproduce
1. Open a new file 'bug.py'
2. Copy the following
```python
from datetime import datetime, datetime, datetime, datetime
from datetime import datetime, datetime, datetime, datetime, datetime
```
- run `M-x flycheck-compile RET python-flake8`. **everything displayed ok**
- move the pointer to `0:0` (line:column) and you see the 4 warnings in the emacs echo area
- now move the pointer to `1:0`. **nothing happens**
- now move the pointer to `1:1`
- then back to `1:0`
3. What happens:
- another window is created in vertical split
- the modeline turns from `(Python FlyC:0/11 AC ElDoc)` into `(Flycheck error messages FlyC)` in both windows
- the content of both windows is still the file I was editing, error messages are not shown.

## Expected behavior
modeline untouched in a window,
the content of the new window should be the flake8 output
I understand there is a limit for the emacs echo area

## Screenshots
https://i.imgur.com/TyGe8rO.png

## System configuration
```
  1 Syntax checkers for buffer bug.py in python-mode:
  2 
  3 First checker to run:
  4 
  5   python-flake8
  6     - may enable:         yes
  7     - executable:         Found at /usr/bin/python
  8     - configuration file: Not found
  9     - `flake8' module:    Found at "/usr/lib/python3.7/site-packages/flake8/__init__.py"
 10 
 11 Checkers that could run if selected:
 12 
 13   python-pycompile  select
 14     - may enable: yes
 15     - executable: Found at /usr/bin/python
 16 
 17 Checkers that are compatible with this mode, but will not run until properly configured:
 18 
 19   python-pylint (disabled)
 20     - may enable:         Automatically disabled!
 21     - executable:         Found at /usr/bin/python
 22     - configuration file: Not found
 23     - `pylint' module:    Missing
 24 
 25   python-mypy (disabled)
 26     - may enable:         Automatically disabled!
 27     - predicate:          t
 28     - executable:         Not found
 29     - configuration file: Not found
 30     - next checkers:      python-flake8
 31 
 32 Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
 33 checkers.
 34 
 35 --------------------
 36 
 37 Flycheck version: 32snapshot (package: 20191028.2247)
 38 Emacs version:    26.3
 39 System:           x86_64-pc-linux-gnu
 40 Window system:    nil
```

Emacs configuration:
- [x] Plain Emacs / Custom configuration (no X)

[Troubleshooting guide]:  https://www.flycheck.org/en/latest/user/troubleshooting.html

**Comment by ssbb** [comment]
Same behavior with lsp-mode for Elm files.

**Comment by fmdkdd** [comment]
Not a bug: we use the built-in `display-message-or-buffer` which opens another window if the message is too long for the echo area, depending on your frame size.  In this case, since there are multiple warnings at point, the message is too long.

You can customize `max-mini-window-height`.

Also, you may be interested in flycheck extensions like [flycheck-inline](https://github.com/flycheck/flycheck-inline).

**Comment by ssbb** [comment]
@fmdkdd Problem not in opening new window for message - thats fine of course. The problem is that it changes current buffer mode to flycheck error mode.

I asked same question yesterday in Elm-lang slack #emacs channel and recorded this video: https://imgur.com/a/GKmHsfE

**Comment by cpitclaudel** [comment]
I can't reproduce this behavior, but it does look like a bug.  This could be linked to https://github.com/flycheck/flycheck/commit/32fb7061ac760f5b8a4800477d7aafed0d8cb3fa , since that added the mode. @rgrinberg , can you reproduce this?

**Comment by rgrinberg** [comment]
Yup. I can reproduce this but not consistently. It seems like in this snippet:
```
(let ((result (display-message-or-buffer (string-join messages "\n\n")
                                               flycheck-error-message-buffer
                                               'not-this-window)))
        (when (window-live-p result)
          (with-current-buffer (window-buffer result)
            (unless (derived-mode-p 'flycheck-error-message-mode)
              (flycheck-error-message-mode)))))
```

`(window-buffer result)` returns the current buffer. I have no idea how.

**Comment by cpitclaudel** [comment]
Any updates on this? I haven't had much time to look into it :/

**Comment by rgrinberg** [comment]
I had a look but not much luck yet. I’m still not sure how the buffer of this window can change from the call to display and the call to get the buffer of this window.

I’ll have another look. If all fails, I guess we can simply double check the buffer name before changing the major mode. We can also scan the buffer list for the appropriately named buffer. But that seems a bit inefficient.

**Comment by manufactory** [comment]
In my opinion this is definitely a bug: I experience a mode change, when editing C files with lots of warnings/errors. Manually changing back to c-mode does not work either.

I'll come up with a minimal example to reproduce as soon as I have time. Hopefully, later today.

**Comment by jdreaver** [comment]
I hate to add onto an issue with just a "me too", but I am experiencing this extremely consistently when running `flycheck` with `haskell-mode`. I will try to make a small repro as well if I have time today.

**Comment by rgrinberg** [comment]
I still don’t quite understand why is display-message-or-buffer returning the wrong window, but I have a workaround for this issue: https://github.com/flycheck/flycheck/pull/1652

Apologies for the inconvenience everyone

**Comment by cpitclaudel** [comment]
I think this was closed in #1652

**Comment by endgame** [comment]
I believe this is still a problem. I am using flycheck with my haskell code (checked with dante), and when I move point onto an error, I get almost identical behaviour to @ssbb 's video above. The only difference is that the mode is preserved, but it seems like when `display-message-or-buffer` opens a new window, it fails to switch to the correct buffer.

This is with GNU Emacs 26.3 and Flycheck 20200610.1809 from MELPA.

I did SFTW before posting, and can't find any discussion suggesting that `display-message-or-buffer` might be broken, but perhaps the root cause is upstream?

**Comment by CastixGitHub** [comment]
@endgame it works for me in emacs 26.3 + flycheck 20200610.1809 from melpa, exactly the same versions. The new opened window have the `special` mode and it's buffer it's named `*Warnings*`

**Comment by endgame** [comment]
@CastixGitHub Thanks for the info, that means my config is the most likely culprit. I'll see if I can slice out a minimal repro, or work out what I've done wrong.
